text/0000-impl-trait-expressions.md

```diff
@@ -0,0 +1,148 @@
+- Feature Name: impl-trait-expressions
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Rust closures allow the programmer to create values of anonymous types which
+implement the `Fn*` traits. This RFC proposes a generalisation of this feature
+to other traits. The syntax looks like this:
+
+```rust
+fn main() {
+    let world = "world";
+    let says_hello_world = impl fmt::Display {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "hello {}", world)
+        }
+    };
+
+    println!("{}", says_hello_world);
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes we need to create a once-off value which implements some trait,
+though having to explicitly declare a type in these situations can be
+unnecessarily painful and noisy. Closures are a good example of how
+this problem can be ameliorated by adding the ability to declare once-off
+values of anonymous types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`impl Trait` expressions allow the user to create a value of an anonymous type which
+implements `Trait`. These expressions behave the same as closures - they
+capture local variables, by either move or reference, and collect them into an
+anonymous struct which implements the required trait.
+
+To better understand the behaviour of `impl Trait`, consider the following code:
+
+```rust
+let y = String::from("hello");
+let foo = move || println!("{}", y);
+```
+
+With this RFC, the above code becomes syntax sugar for:
+
+```rust
+let y = String::from("hello");
+let foo = move impl FnOnce<()> {
+    type Output = ();
+
+    extern "rust-call" fn call_once(self, args: ()) {
+        println!("{}", y);
+    }
+};
+```
+
+Which, in turn, is syntax sugar for:
+
+```rust
+let y = String::from("hello");
+
+struct MyCoolAnonType {
+    y: String,
+}
+
+impl FnOnce<()> for MyCoolAnonType {
+    type Output = ();
+
+    extern "rust-call" fn call_once(self, args: ()) {
+        println!("{}", self.y);
+    }
+}
+
+let foo = MyCoolAnonType { y };
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature is fully described in the guide-level explanation. As this is a
+generalisation of the existing closure syntax I suspect that the implementation
+would be fairly straight-forward.
+
```

Comment:
As for the ambiguity aforementioned, consider:

```rust
fn foo() {
    struct X;
    
    impl X {
        fn foo() {}
    };
}
```
This compiles today and `impl X { ... }` is an `item` followed by the empty statement `;`.

According to your RFC however `impl X { ... }` is an expression which is made into a statement by the following `;`.

This is not an insurmountable challenge but you'll need to think about how to deal with it.

Reply:
cc @petrochenkov @eddyb @qmx ^ ideas?